### ISSUE TITLE
l_slope is a bit slow

### ISSUE THREAD

**Issue by njtierney** [njtierney/brolgar#6]
It would be great to improve the speed of l_slope

``` r
l_slope_old <- function(df, id, formula){
  l <- split(df, df[[id]])
  sl <- purrr::map(l, ~ eval(substitute(lm(formula, data = .)))) %>%
    purrr::map_dfr( ~ as.data.frame(t(as.matrix(coef(
      .
    ))))) %>%
    dplyr::mutate(id = as.integer(names(l))) %>%
    dplyr::rename_all( ~ c("intercept", "slope", "id")) %>%
    dplyr::select(id, intercept, slope) %>%
    tibble::as_tibble()
  return(sl)
}

library(brolgar)

bm1 <- bench::mark(old = l_slope_old(wages, "id", "lnw~exper"),
                   new = l_slope(wages, id, lnw~exper))
#> Warning: Some expressions had a GC in every iteration; so filtering is
#> disabled.

summary(bm1)
#> Warning: Some expressions had a GC in every iteration; so filtering is
#> disabled.
#> # A tibble: 2 x 10
#>   expression      min     mean   median      max `itr/sec` mem_alloc  n_gc
#>   <chr>      <bch:tm> <bch:tm> <bch:tm> <bch:tm>     <dbl> <bch:byt> <dbl>
#> 1 old        748.71ms 748.71ms 748.71ms 748.71ms     1.34     7.25MB    16
#> 2 new           2.45s    2.45s    2.45s    2.45s     0.408   11.93MB    67
#> # … with 2 more variables: n_itr <int>, total_time <bch:tm>
summary(bm1, relative = TRUE)
#> Warning: Some expressions had a GC in every iteration; so filtering is
#> disabled.
#> # A tibble: 2 x 10
#>   expression   min  mean median   max `itr/sec` mem_alloc  n_gc n_itr
#>   <chr>      <dbl> <dbl>  <dbl> <dbl>     <dbl>     <dbl> <dbl> <dbl>
#> 1 old         1     1      1     1         3.27      1     1        1
#> 2 new         3.27  3.27   3.27  3.27      1         1.65  4.19     1
#> # … with 1 more variable: total_time <dbl>
```

<sup>Created on 2019-04-02 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by njtierney** [comment]
I should try out using `furrr` to make this faster

**Comment by sa-lee** [comment]
I had a go at speeding this up... this version just uses summarise, avoids broom and dplyr renaming functions. It can also handle multiple variables on the RHS as per #12. Let me know what you think.
``` r
l_slope_new <- function(data, id, formula) {
  
  quo_id <- rlang::enquo(id)
  quo_formula <- rlang::enquo(formula)
  f_rhs_vars <- all.vars(rlang::f_rhs(formula))
  coef_tbl_vars <- c(rlang::as_label(quo_id),
                     "l_intercept", 
                     paste0("l_slope_", f_rhs_vars))
  
  data %>%
    dplyr::group_by(!!quo_id) %>%
    dplyr::summarise(
      coef_tbl = list(as.data.frame(t(coef(lm(!!quo_formula)))))
    ) %>%
    tidyr::unnest() %>%
    rlang::set_names(coef_tbl_vars)
}

l_slope_old <- function(df, id, formula){
  l <- split(df, df[[id]])
  sl <- purrr::map(l, ~ eval(substitute(lm(formula, data = .)))) %>%
    purrr::map_dfr( ~ as.data.frame(t(as.matrix(coef(
      .
    ))))) %>%
    dplyr::mutate(id = as.integer(names(l))) %>%
    dplyr::rename_all( ~ c("intercept", "slope", "id")) %>%
    dplyr::select(id, intercept, slope) %>%
    tibble::as_tibble()
  return(sl)
}


library(brolgar)

bm1 <- bench::mark(old = l_slope_old(wages, "id", "lnw~exper"),
                   cur = l_slope(wages, id, lnw~exper),
                   new = l_slope_new(wages, id, lnw~exper),
                   check = FALSE
                   )
#> Warning: Some expressions had a GC in every iteration; so filtering is
#> disabled.

summary(bm1)
#> Warning: Some expressions had a GC in every iteration; so filtering is
#> disabled.
#> # A tibble: 3 x 10
#>   expression     min    mean  median     max `itr/sec` mem_alloc  n_gc
#>   <chr>      <bch:t> <bch:t> <bch:t> <bch:t>     <dbl> <bch:byt> <dbl>
#> 1 old        839.1ms 839.1ms 839.1ms 839.1ms     1.19     6.93MB    19
#> 2 cur          2.01s   2.01s   2.01s   2.01s     0.498   11.89MB    21
#> 3 new        576.2ms 576.2ms 576.2ms 576.2ms     1.74     4.75MB     6
#> # … with 2 more variables: n_itr <int>, total_time <bch:tm>

summary(bm1, relative = TRUE)
#> Warning: Some expressions had a GC in every iteration; so filtering is
#> disabled.
#> # A tibble: 3 x 10
#>   expression   min  mean median   max `itr/sec` mem_alloc  n_gc n_itr
#>   <chr>      <dbl> <dbl>  <dbl> <dbl>     <dbl>     <dbl> <dbl> <dbl>
#> 1 old         1.46  1.46   1.46  1.46      2.39      1.46  3.17     1
#> 2 cur         3.48  3.48   3.48  3.48      1         2.50  3.5      1
#> 3 new         1     1      1     1         3.48      1     1        1
#> # … with 1 more variable: total_time <dbl>
```

<sup>Created on 2019-04-10 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by njtierney** [comment]
Nice! And two birds with one stone! This is both different and more elegant than what I was thinking. Thanks!

Do you think you could submit this as a PR, and add some tests to cover the fact that multiple things on the RHS work now? Also no worries if you don't want to add tests, I can do that if you prefer.

**Comment by sa-lee** [comment]
sure sounds good! I'll get on to that in a jiffy. feel free to fix any coding style things etc.

**Comment by njtierney** [comment]
Excellent!

And sure, once you've submitted the PR I might make a few changes. :)

**Comment by njtierney** [comment]
closed in #21
